smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java

```diff
@@ -0,0 +1,390 @@
+     */
+    public static TraceFile createFromNode(Node value) {
+        ObjectNode node = value.expectObjectNode();
+        Builder builder = builder()
+                .smithyTrace(node.expectStringMember(SMITHY_TRACE_TEXT).getValue())
+                .artifact(ArtifactMetadata.createFromNode(node.expectObjectMember(ARTIFACT_TEXT)));
+
+        //parse shapes
+        Map<StringNode, Node> shapeMap = node.expectObjectMember(SHAPES_TEXT).getMembers();
+        for (Map.Entry<StringNode, Node> entry : shapeMap.entrySet()) {
+            for (Node linkNode : (entry.getValue().expectArrayNode()).getElements()) {
+                builder.addShapeLink(entry.getKey().getValue(), ShapeLink.createFromNode(linkNode));
+            }
+        }
+
+        //parse definitions
+        if (node.containsMember(DEFINITIONS_TEXT)) {
+            builder.definitions(ArtifactDefinitions.createFromNode(node.expectObjectMember(DEFINITIONS_TEXT)));
+        }
+
+        return builder.build();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Writes the TraceFile JSON ObjectNode
+     * to construct a trace file.
+     *
+     * @param fileName absolute or relative path to write the trace file
+     * @throws IOException if there is an error writing to fileName
+     */
+    public void writeTraceFile(String fileName) throws IOException {
+        Writer writer = new OutputStreamWriter(new FileOutputStream(new File(fileName)), StandardCharsets.UTF_8);
+        PrintWriter pw = new PrintWriter(writer);
+        pw.print(Node.prettyPrintJson(toNode(), "  "));
+        pw.close();
+    }
+
+    /**
+     * Converts TraceFile instance variables into an
+     * ObjectNode.
+     *
+     * @return ObjectNode representation of a TraceFile.
+     */
+    @Override
+    public ObjectNode toNode() {
+        //constructing shapes ObjectNode map
+        ObjectNode.Builder shapesBuilder = ObjectNode.objectNodeBuilder();
+        for (Map.Entry<ShapeId, List<ShapeLink>> entry : shapes.entrySet()) {
+            shapesBuilder.withMember(entry.getKey().toString(),
+                    entry.getValue().stream().map(ShapeLink::toNode).collect(ArrayNode.collect()));
+        }
+
+        //returning ObjectNode for TraceFile
+        return ObjectNode.objectNodeBuilder()
+                .withMember(SMITHY_TRACE_TEXT, smithyTrace)
+                .withMember(ARTIFACT_TEXT, artifactMetadata)
+                .withOptionalMember(DEFINITIONS_TEXT, getArtifactDefinitions())
+                .withMember(SHAPES_TEXT, shapesBuilder.build())
+                .build();
+    }
+
+    /**
+     * Finds invalid types and tags and either removes them or throws an error depending on
+     * whether toThrow is true or false.
+     *
+     * @throws ExpectationNotMetException if a type or tag in shapes is not in definitions.
+     */
+    public void validateTypesAndTags() {
```

Comment:
This method needs a little better documentation around its precondition and use case.